Add ExponentialHistogram to Metrics data model (#1935)
open-telemetry/opentelemetry-specification: specification/metrics/datamodel.md

  * [Sums](#sums)
  * [Gauge](#gauge)
  * [Histogram](#histogram)
  * [ExponentialHistogram](#exponentialhistogram)

>     + [Exponential scale](#exponential-scale)

Author: Done!

and a new aggregation occurs. Cumulative, on the other hand, continues to
aggregate events, resetting with the use of a new start time.

> ### ExponentialHistogram

Reviewer: This should have a stability label.

Author: Fixed. I marked this "experimental" is that right?

and a new aggregation occurs. Cumulative, on the other hand, continues to
aggregate events, resetting with the use of a new start time.

### ExponentialHistogram

**Status**: [Experimental](../document-status.md)

[ExponentialHistogram](https://github.com/open-telemetry/opentelemetry-proto/blob/cfbf9357c03bf4ac150a3ab3bcbe4cc4ed087362/opentelemetry/proto/metrics/v1/metrics.proto#L222)
data points are an alternate representation to the
[Histogram](#histogram) data point, used to convey a population of
recorded measurements in a compressed format.  ExponentialHistogram
compresses bucket boundaries using an exponential formula, making it

> suitable for conveying high-resolution data using a relatively large

Reviewer 2: sounds good

subsetting".  Buckets of an exponential Histogram with a given scale
map exactly into buckets of exponential Histograms with lesser scales,
which allows consumers to automatically lower the resolution of a
histogram (i.e., downscale) without introducing error.

#### Exponential buckets

The ExponentialHistogram bucket identified by `index`, a signed
integer, represents values in the population that are greater than or
equal to `base**index` and less than `base**(index+1)`.

The positive and negative ranges of the histogram are expressed
separately.  Negative values are mapped by their absolute value
into the negative range using the same scale as the positive range.

Each range of the ExponentialHistogram data point uses a dense
representation of the buckets, where a range of buckets is expressed
as a single `offset` value, a signed integer, and an array of count
values, where array element `i` represents the bucket count for bucket
index `offset+i`.

For a given range, positive or negative:

- Bucket index `0` counts measurements in the range `[1, base)`
- Positive indexes correspond with absolute values greater or equal to `base`

> - Negative indexes correspond with absolute values less than 1

Reviewer 3: Question: Assuming bucket counts are represented by a list, how would you expect implementations to represent bucket counts with a negative index? Would this be achieved by adjusting the `offset` according to the lowest measured value?

Reviewer 2: "adjusting the offset according to the lowest measured value" is certainly one way. But it involves copying the array when offset changes. Another way is [WindowedCounterArray](https://github.com/newrelic-experimental/newrelic-sketch-java/blob/main/src/main/java/com/newrelic/nrsketch/WindowedCounterArray.java), used in [NrSketch](https://github.com/newrelic-experimental/newrelic-sketch-java#windowedcounterarray). It automatically adjusts to the incoming data's index range without shifting or copying the array.

- There are `2**scale` buckets between successive powers of 2.

For example, with `scale=3` there are `2**3` buckets between 1 and 2.
Note that the lower boundary for bucket index 4 in a `scale=3`
histogram maps into the lower boundary for bucket index 2 in a
`scale=2` histogram and maps into the lower boundary for bucket index

> 1 (i.e., the `base`) in a `scale=1` histogram.

Reviewer 2: maybe add "This is an example of perfect subsetting"

| --                     | --                | --                           |
| 0                      | 1                 | 2**(0/8)                     |
| 1                      | 1.090507732665258 | 2**(1/8)                     |
| 2                      | 1.189207115002721 | 2**(2/8), 2**(1/4)           |
| 3                      | 1.296839554651010 | 2**(3/8)                     |
| 4                      | 1.414213562373095 | 2**(4/8), 2**(2/4), 2**(1/2) |
| 5                      | 1.542210825407941 | 2**(5/8)                     |
| 6                      | 1.681792830507429 | 2**(6/8)                     |
| 7                      | 1.834008086409343 | 2**(7/8)                     |

#### Zero count

The ExponentialHistogram contains a special `zero_count` field
containing the count of values that are either exactly zero or within
the region considered zero by the instrumentation at the tolerated
level of precision.  This bucket stores values that cannot be
expressed using the standard exponential formula as well as values
that have been rounded to zero.

#### Producer expectations

The ExponentialHistogram design makes it possible to express values

> that are too large or small to be represented in computer hardware.

Reviewer 2: It's more like "computer software". Suggest changing "computer hardware" to "the 64 bit "double" floating point format"

Certain values for `scale`, while meaningful, are not necessarily
useful.

The range of data represented by an ExponentialHistogram determines
which scales can be usefully applied.  Therefore, producers SHOULD

> ensure that bucket indices are within the range of a signed 64-bit

Reviewer 2: you mean "32 bit (because we currently use 32 bit for index offset in protocol)"? 
The issue here is that even with the 32bit offset, a producer can still produce an index beyond 32 bit. For example, when offset is close to 32 bit max, then offset + arrayIndex will still go above 32 bit max. 

Shall we say something about such index overflow in consumer section?

Author: I wrote 
```
Regardless of scale, producers
SHOULD ensure that the index of any encoded bucket falls within the
range of a signed 32-bit integer.  This recommendation is applied to
limit the width of integers used in standard processing pipelines such
as the OpenTelemetry collector.  The wire-level protocol could be
extended for 64-bit bucket indices in a future release.
```

> integer by changing scale.

Author: @yzhuge reviewing https://github.com/newrelic-experimental/newrelic-sketch-java/blob/1ce245713603d61ba3a4510f6df930a5479cd3f6/src/main/java/com/newrelic/nrsketch/indexer/ScaledExpIndexer.java#L26 -- would you say 52bits is a better limit for the index?

Reviewer 2: starting at https://github.com/newrelic-experimental/newrelic-sketch-java/blob/main/src/test/java/com/newrelic/nrsketch/indexer/BucketIndexerTest.java#L242, I have scale accuracy tests 
 testLogIndexerScales, testSubBucketLogIndexerScales, testSubBucketLookupIndexerScales, and testExponentIndexerScales. Summary here:

64 bit index is needed to cover double min (including subnormals) to max at scale -11 to 52. At scale -11, there are just two buckets, 0 for >= 1 values, and -1 for < 1 values. At scale 52, index "mantissa" bits == value mantissa bits. Higher scale is pointless (they will just create many empty buckets). 

Log indexer can go up to scale 42, with index off by at most 1, which is expected due to inherent floating point inaccuracy. Beyond that, it will be off by 2 or more.

Some indexers, like SubBucketLogIndexer can go to scale 52 with off by 1. SubBucket Lookup indexer of course is limited by memory size. The subbucket indexers get better accuracy because they only use log on the mantissa, which has limited range of 1 to 2.

Author: Thank you!

Producers MAY use a built-in logarithm function to calculate the
bucket index of a value.  The use of a built-in logarithm function
could lead to results that differ from the bucket index that would be
computed using arbitrary precision or a lookup table, however
producers are not required to perform an exact computation.  As a
result, ExponentialHistogram exemplars could map into buckets with

> zero count.  Instead, we expect to find such values counted in the

Reviewer 2: "instead" seems extraneous.

adjacent bucket.

#### Consumer expectations

ExponentialHistogram buckets are expected to map into numbers that can be
represented using normalized IEEE 754 double-width floating point

> values (i.e., subnormal values are excluded).  Consumers SHOULD reject

Reviewer 2: Subnormal exclusion here seems odd. We didn't talk about it anywhere else in this doc, most importantly, not in the producer section. I'd like to remove this exclusion. If we have to say something about subnormals, we can say:
Consumer may choose to round subnormals into the zero_count bucket, or reject such numbers, if it is incapable of handling subnormal numbers.

Author: I removed it. I believe the mapping functions are well-defined for subnormals, I just don't want to think about them 😁

| Scale | Base    | Expression  |
| --    | --      | --          |
| 10    | 1.00068 | 2**(1/1024) |
| 9     | 1.00135 | 2**(1/512)  |
| 8     | 1.00271 | 2**(1/256)  |
| 7     | 1.00543 | 2**(1/128)  |
| 6     | 1.01089 | 2**(1/64)   |
| 5     | 1.02190 | 2**(1/32)   |
| 4     | 1.04427 | 2**(1/16)   |
| 3     | 1.09051 | 2**(1/8)    |
| 2     | 1.18921 | 2**(1/4)    |
| 1     | 1.41421 | 2**(1/2)    |
| 0     | 2       | 2**1        |
| -1    | 4       | 2**2        |
| -2    | 16      | 2**4        |
| -3    | 256     | 2**8        |
| -4    | 65536   | 2**16       |

An important property of this design is described as "perfect
subsetting".  Buckets of an exponential Histogram with a given scale
map exactly into buckets of exponential Histograms with lesser scales,
which allows consumers to lower the resolution of a histogram (i.e.,
downscale) without introducing error.

#### Exponential buckets

The ExponentialHistogram bucket identified by `index`, a signed
integer, represents values in the population that are greater than or
equal to `base**index` and less than `base**(index+1)`.  Note that the
ExponentialHistogram specifies a lower-inclusive bound while the

> explicit-boundary Histogram specifies an upper-inclusive bound.

Reviewer: Glad you call this one out!

The ExponentialHistogram design makes it possible to express values
that are too large or small to be represented in the 64 bit "double"
floating point format.  Certain values for `scale`, while meaningful,
are not necessarily useful.

The range of data represented by an ExponentialHistogram determines
which scales can be usefully applied.  Regardless of scale, producers
SHOULD ensure that the index of any encoded bucket falls within the
range of a signed 32-bit integer.  This recommendation is applied to
limit the width of integers used in standard processing pipelines such
as the OpenTelemetry collector.  The wire-level protocol could be
extended for 64-bit bucket indices in a future release.

Producers MAY use a built-in logarithm function to calculate the
bucket index of a value.  The use of a built-in logarithm function
could lead to results that differ from the bucket index that would be
computed using arbitrary precision or a lookup table, however
producers are not required to perform an exact computation.  As a
result, ExponentialHistogram exemplars could map into buckets with
zero count.  We expect to find such values counted in the adjacent
bucket.

> 

Reviewer 4: I would prefer "in adjacent buckets" over "in the adjacent bucket" as values may also be mapped to non-direct neighbors.

Author: Done.

   representation `1._significand_ * 2**_exponent_`.  Normal IEEE 754
   double-width floating point values have indices in the range
   [-1022,+1023] and subnormal values have indices in the range
   [-1074,-1023].  This may be written as:

```golang
// GetExponent extracts the normalized base-2 fractional exponent.
// Let the value be represented as `1.significand x 2**exponent`,
// this returns `exponent`.  Not defined for 0, Inf, or NaN values.
func GetExponent(value float64) int32 {
	rawBits := math.Float64bits(value)
	rawExponent := (int64(rawBits) & ExponentMask) >> SignificandWidth
	rawSignificand := rawBits & SignificandMask
	if rawExponent == 0 {
		// Handle subnormal values: rawSignificand cannot be zero
		// unless value is zero.
		rawExponent -= int64(bits.LeadingZeros64(rawSignificand) - 12)
	}
	return int32(rawExponent - ExponentBias)
}
```

2. For negative scales, the index of a value equals the normalized
   base-2 exponent (as by `GetExponent()` above) shifted to the right
   by `-scale`.  Note that because of sign extension, this shift performs
   correct rounding for the negative indices.  This may be written as:
   
```golang

>   return GetExponent(value) << -scale

Reviewer 2: you mean ">>"?

Author: Yes, thanks.

```
   
3. For positive scales, use of the built-in natural logarithm
   function.  A multiplicative factor equal to `2**scale / ln(2)`
   proves useful (where `ln()` is the natural logarithm), for example:
   
```golang

>     scaleFactor := math.Log2E * math.Exp2(1<<scale)

Reviewer 2: should this be "math.Exp2(scale)"?
And log method is generally applicable on 0 and negative scales too, though less efficient. For simplicity, an implementation may just do a one size fits all log method, especially when it expects that 0 and negative scales are used rarely.

Author: Fixed. Thanks.

const (
    // SignificandWidth is the size of an IEEE 754 double-precision
    // floating-point significand.
    SignificandWidth = 52
    // ExponentWidth is the size of an IEEE 754 double-precision
    // floating-point exponent.
    ExponentWidth = 11

    // SignificandMask is the mask for the significand of an IEEE 754
    // double-precision floating-point value: 0xFFFFFFFFFFFFF.
    SignificandMask = 1<<SignificandWidth - 1

    // ExponentBias is the exponent bias specified for encoding
    // the IEEE 754 double-precision floating point exponent: 1023.
    ExponentBias = 1<<(ExponentWidth-1) - 1

    // ExponentMask are set to 1 for the bits of an IEEE 754
    // floating point exponent: 0x7FF0000000000000.
    ExponentMask = ((1 << ExponentWidth) - 1) << SignificandWidth
)
```

The following choices of mapping function have been validated through
reference implementations.

##### Scale zero: extract the exponent

For scale zero, the index of a value equals its normalized base-2
exponent, meaning the value of _exponent_ in the base-2 fractional
representation `1._significand_ * 2**_exponent_`.  Normal IEEE 754
double-width floating point values have indices in the range

> `[-1022,+1023]` and subnormal values have indices in the range

Reviewer 5: ```suggestion
`[-1022, +1023]` and subnormal values have indices in the range
```

> `[-1074,-1023]`.  This may be written as:

Reviewer 5: ```suggestion
`[-1074, -1023]`.  This may be written as:
```

lookup table with `O(2**scale)` entries.

##### Producer Recommendations

At the lowest or highest end of the 64 bit IEEE floating point, a
bucket's range may only be partially representable by the floating
point number format.  When mapping a number in these buckets, a
producer may correctly return the index of such a partially
representable bucket.  This is considered a normal condition.

For positive scales, the logarithm method is preferred because it
requires very little code, is easy to validate and is nearly as fast
and accurate as the lookup table approach.  For zero scale and
negative scales, directly calculating the index from the
floating-point representation is more efficient.

The use of a built-in logarithm function could lead to results that
differ from the bucket index that would be computed using arbitrary
precision or a lookup table, however producers are not required to
perform an exact computation.  As a result, ExponentialHistogram
exemplars could map into buckets with zero count.  We expect to find
such values counted in the adjacent buckets.

#### Consumer Expectations

ExponentialHistogram bucket indices are expected to map into buckets

> where both the uppwer and lower boundaries that can be represented

Reviewer 6: ```suggestion
where both the upper and lower boundaries can be represented
```

##### Producer recommendations

> Regardless of scale or mapping technique, it can be difficult to

Author: Done -- approximately. I put your final sentence into the Consumer section and reworded slightly. IIUC, this can be stated as "Consumers MAY round partially representable bucket indices up or down to the adjacent fully-representable bucket index."

> Regardless of scale or mapping technique, it can be difficult to

Reviewer 2: I meant consumer keeping the partial bucket, but capping the bound to a representable number, not folding the bucket to another bucket. For example, if last bucket is logically 90 to 100, but the format can only represent up to 95, then the consumer shrinks the bucket to 90 to 95. The returned bound is typically used for charting or percentile calculation. With the capping, the result is still logically correct. No "rounding" error is introduced. Rounding to a full bucket is an alternative, but it introduces rounding error.
